cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java

```diff
@@ -79,19 +96,59 @@ public void initialize() throws Exception {
     // because it holds one thread per receiver, or one core in distributed mode.
     // so... we have to set this hacky master variable based on the isUnitTest setting in the config
     Map<String, String> programProperties = context.getSpecification().getProperties();
-    String extraOpts = programProperties.get("cask.hydrator.extra.opts");
+    String extraOpts = programProperties.get(EXTRA_OPTS);
     if (extraOpts != null && !extraOpts.isEmpty()) {
       sparkConf.set("spark.driver.extraJavaOptions", extraOpts);
       sparkConf.set("spark.executor.extraJavaOptions", extraOpts);
     }
-    Integer numSources = Integer.valueOf(programProperties.get("cask.hydrator.num.sources"));
+    Integer numSources = Integer.valueOf(programProperties.get(NUM_SOURCES));
     // without this, stopping will hang on machines with few cores.
     sparkConf.set("spark.rpc.netty.dispatcher.numThreads", String.valueOf(numSources + 2));
-    Boolean isUnitTest = Boolean.valueOf(programProperties.get("cask.hydrator.is.unit.test"));
+    Boolean isUnitTest = Boolean.valueOf(programProperties.get(IS_UNIT_TEST));
     if (isUnitTest) {
       sparkConf.setMaster(String.format("local[%d]", numSources + 1));
     }
     context.setSparkConf(sparkConf);
+
+    if (!checkpointsDisabled) {
+      // Each pipeline has its own checkpoint directory within the checkpoint fileset.
+      // Ideally, when a pipeline is deleted, we would be able to delete that checkpoint directory.
+      // This is because we don't want another pipeline created with the same name to pick up the old checkpoint.
+      // Since CDAP has no way to run application logic on deletion, we instead generate a unique pipeline id
+      // and use that as the checkpoint directory as a subdirectory inside the pipeline name directory.
+      // On start, we check for any other pipeline ids for that pipeline name, and delete them if they exist.
```

Comment:
That's a good question. One option would be to rename this property from 'pipelineid' to 'checkpointdir', and to make checkpointdir a property in the application config. If one is not given, the app will generate one as a UUID during configure time. Otherwise, it will use whatever the user supplies. 

If a user wants to add a new version of the pipeline, they can look up the checkpointdir from the previous version since it is just an application property, and set it in the application config.